packages/nimble-components/src/table/specs/data-hierarchy-hld.md

```diff
@@ -0,0 +1,121 @@
+2. After providing the current data to the `Table` via the `setData` method, all rows that have a value of `true` in the field specified by the `hasChildrenFieldName` attribute will display and expand/collapse button.
+3. Users must register a handler for the `row-expand-toggle` event on the `Table` instance, and will receive that event upon clicking the expand/collapse button.
+4. The details of the handled event will include the id for the row that was expanded.
+5. The user must then create a set of data with rows where they provide a field with the name specified by the `parendIdFieldName` attribute that has the value of the `recordId` value supplied in the event details.
+6. The user then sets the data on the `Table` again with the `setData` method.
+
+Notes:
+
+-   It is up to the user to manage whether or not the children of a row has already been loaded in order to avoid recreation of their data and calling `setData` on the `Table` unnecessarily.
+
+### Translating flat list to Tanstack-understandable hierarchy
+
+Tanstack provides APIs for us to implement that allow it to return the rows in a hierarchical fashion where child rows are provided as a property on a row called `subRows`. For this to work as expected it is required that a flat list of data (that contains implicit hierarchy) be transformed into a hierarchical data structure.
+
+A third party library called [`performant-array-to-tree`](https://www.npmjs.com/package/performant-array-to-tree) offers an easy, and as the package says, performant means of doing this (in O(n) time). It comes with an MIT license, and is apparently fully tested, so it seems like it would be suitable to use for this purpose.
+
+### Group rows with data hierarchy
+
+Tanstack allows limited support of data modeling where there is both grouping and parent-child row relationships. Essentially, when there is grouping present, only rows at the top-level of a hierarchical set of data will be grouped. All further child rows under the top-level parent row will not be grouped.
+
+I see no reason to explicitly disable this behavior. One behavior we will need to ensure, however, is that the count value we display in the group row, is _only_ the number of immediate children in the group row. It would be odd for the number to change just because a row was expanded (such as in the lazy loading case).
+
+#### Managing expanded state
+
+Currently, the `Table` defaults to expanding all rows by setting its Tanstack expanded state to the singleton `true`. This means that all group rows will be expanded by default. The Nimble `Table` also provides an implementation of `getIsRowExpanded` which is consumed by `TanStack` that will ultimately result in the per-row state of whether the row is expanded. It is in this implementation, where even if the Tanstack state is set to `true`, we can mark all parent rows, by default, as collapsed, which is required for the lazy-loading scenario.
```

Comment:
I mentioned this in a section above, but I think we're going to need configuration for "expand parent rows automatically". There are definitely cases where we expand the parent rows by default, and I don't think we'll want to change that behavior.

Reply:
I'll leave this thread open as well, and update this section as needed pending any decisions we make.